src/LDATaskS3.cpp

```diff
@@ -0,0 +1,385 @@
+  if (end > train_range.second) {
+    end = train_range.second;
+  }
+  int cur_train_idx = start;
+
+  bool printed_rate = false;
+  int count = 0;
+  auto start_time = get_time_ms();
+
+  int update_bucket = 0;
+  int benchmark_time = 5, full_iteration = 0;
+  uint32_t to_receive_size, uncompressed_size;
+  int total_sampled_tokens = 0, total_sampled_docs = 0;
+  uint64_t version = 1;
+  upload_lock_indicators.resize(end - start, -1);
+
+  // Load the first LDAStatistics from S3
+  auto start_time_benchmark = get_time_ms();
+  // s3_initialize_aws();
+  std::shared_ptr<S3Client> s3_client = std::make_shared<S3Client>();
+  std::string obj_id_str =
+      std::to_string(hash_f(std::to_string(cur_train_idx).c_str())) + "-LDA";
+  std::ostringstream* s3_obj =
+      s3_client->s3_get_object_ptr(obj_id_str, config.get_s3_bucket());
+  time_download += (get_time_ms() - start_time_benchmark) / 1000.0;
+
+  // create a LDAModel object with local statistics from S3
+  std::unique_ptr<LDAModel> model;
+  uint64_t to_send_size;
+  model.reset(new LDAModel(s3_obj->str().c_str()));
+  delete s3_obj;
+
+  // load the pre-cached indices from server
+  char* slice_indices_mem = psint->get_slices_indices(cur_train_idx);
+  load_serialized_indices(slice_indices_mem);
+  delete slice_indices_mem;
+
+  int cur = 0, num_runs = 1;
+  std::cout << "[WORKER] starting loop" << std::endl;
+
+  while (1) {
+#ifdef DEBUG
+    std::cout << get_time_us() << " [WORKER] running phase 1" << std::endl;
+#endif
+    // only True if current LDAStatistics has been sampled with
+    // more than (num_runs) number of word slices
+    if (cur >= num_runs) {
+      // update_bucket = cur_train_idx;
+      total_sampled_docs += model->get_ndt_size();
+
+      // only send to S3 if the next LDAStatistics is not the current one
+      if (end != start + 1) {
+        uint64_t size_temp;
+        char* mem_to_send = model->serialize_to_S3(size_temp);
+        upload_wih_bucket_id_fn(mem_to_send, size_temp,
+                                upload_lock_indicators[0], cur_train_idx);
+      }
+
+      // compute the loglikelihood for current LDAStatistics and
+      // send to server
+      double ll_to_send = model->compute_ll_ndt();
+      psint->update_ll_ndt(cur_train_idx, ll_to_send);
+
+      // Get the next LDAStatistics from S3
+      cur_train_idx += 1;
+
+      if (cur_train_idx == end) {
+        cur_train_idx = start;
+        full_iteration += 1;
+
+        // The early exit is added in case that
+        // worker expires even if it hasn't stored the
+        // updated LDAStatistics to S3
+        auto elapsed_ms = get_time_ms() - start_time;
+        float elapsed_sec = elapsed_ms / 1000.0;
+
+        // estimate average time to sample for one iteration
+        float est_time_one_iter =
+            ((get_time_ms() - start_time) / 1000.) / full_iteration;
+
+        if (elapsed_sec > (lambda_time_out - est_time_one_iter - 10.)) {
```

Comment:
Too much code here. Would put this code below in a separate function.

Reply:
Code refactoring has been done. Please let me know if it remains messy.